src/of-type.ts

```diff
@@ -23,26 +23,14 @@ import { castArray } from './utils'
  */
 export function ofType<
   TSource extends AnyAction,
-  TActionCreator extends ActionCreator<TSource>,
->(
-  actionCreators: TActionCreator | ReadonlyArray<TActionCreator>
-): OperatorFunction<TSource, ReturnType<TActionCreator>>
-export function ofType<TSource extends AnyAction, TAction extends TSource>(
-  actions: TAction | ReadonlyArray<TAction>
-): OperatorFunction<TSource, TAction>
-export function ofType<
-  TSource extends AnyAction,
-  TType extends TSource['type']
->(
-  types: TType | ReadonlyArray<TType>
-): OperatorFunction<
-  TSource,
-  TSource extends Action<infer T> ? (T extends TType ? TSource : never) : never
->
-export function ofType(keys: ActionCreator<AnyAction> | AnyAction | string) {
-  const types = castArray(keys).map(key =>
-    typeof key === 'string' ? key : getType(key)
-  )
+  TKey extends string | AnyAction | ActionCreator<AnyAction>,
```

Comment:
`TKey` should extend `TSource` for each sum type.

Reply:
Something like `TSource['type'] | TSource | ActionCreator<TSource>`.

Reply:
DTS test error


```
  ● ofType › of(a(), b(), c()).pipe(ofType(['a', 'b'])) › (type) should match snapshot

  expect(received).toMatchSnapshot()

  Snapshot name: `ofType of(a(), b(), c()).pipe(ofType(['a', 'b'])) (type) should match snapshot 1`

  Snapshot: "Observable<{ type: \"a\"; } | { type: \"b\"; }>"
  Received: "Observable<{ type: \"a\"; } | { type: \"b\"; } | { type: \"c\"; }>"

      27 |
      28 | // @dts-jest:pass:snap
    > 29 | of(a(), b(), c()).pipe(ofType(['a', 'b']))
         |                                                                                                       ^
      30 |
      31 | // @dts-jest:pass:snap
      32 | of(a(), b(), c(), d()).pipe(ofType(['a', b(), c]))

      at Object.<anonymous> (src/__tests__/of-type.dts.spec.ts:29:345)
```

```
  ● ofType › of(a(), b(), c(), d()).pipe(ofType(['a', b(), c])) › (type) should match snapshot

  expect(received).toMatchSnapshot()

  Snapshot name: `ofType of(a(), b(), c(), d()).pipe(ofType(['a', b(), c])) (type) should match snapshot 1`

  Snapshot: "Observable<{ type: \"a\"; } | { type: \"b\"; } | { type: \"c\"; }>"
  Received: "Observable<{ type: \"a\"; } | { type: \"b\"; } | { type: \"c\"; } | { type: \"d\"; }>"

      30 |
      31 | // @dts-jest:pass:snap
    > 32 | of(a(), b(), c(), d()).pipe(ofType(['a', b(), c]))
         |                                                                                                                         ^
      33 |

      at Object.<anonymous> (src/__tests__/of-type.dts.spec.ts:32:353)
```

Reply:
Ah, in fact, it's right (There is no need to filter actions that won't receive by a stream), but as you can see in both cases that tests failed we need to cast array to be readonly.

Reply:
> Ah, in fact, it's right (There is no need to filter actions that won't receive by a stream)

I'm not good at English.
Sorry, Can you tell me what the "it" here refers to?

> but as you can see in both cases that tests failed we need to cast array to be readonly.

Also, Can you tell me what the "both cases" here refers to?
In the case of `TKey extends string | AnyAction | ActionCreator<AnyAction>`, there is no need to cast array to be readonly.

Reply:
Sorry for the ambiguity.

> Sorry, Can you tell me what the "it" here refers to?

I meant `TSource['type'] | TSource | ActionCreator<TSource>` is right type.

> Also, Can you tell me what the "both cases" here refers to?

I meant test cases that have been failed.

> In the case of TKey extends string | AnyAction | ActionCreator<AnyAction>, test succeed and there is no need to cast array to be readonly.

It does but there is an issue. Let me explain.
Imagine your `TSource` is `{ type: 'a' } | { type: 'b' }`.
If we don't restrict `TKey` to `TSource` following code won't throw any error:
```ts
of(a(), b()).pipe(ofType(c))
```
Although at runtime there won't be any problem with the above scenario, at compile-time we know that `TSource` will never be `c` so if we restrict `TKey` based on `TSource` the above code will throw an error at compile-time.

Reply:
Your explanation really helped!
Thank you very much!

> There is no need to filter actions that won't receive by a stream

I understand about the above.
As you explained, we need to restrict `TKey` based on `TSource`.

Reply:
completed! https://github.com/thebrodmann/deox/pull/133/commits/d3f2f274988a0a2cafad5ce3a61beeb6f31d06da